### PR TITLE
Support InternVL2.5

### PR DESCRIPTION
**Pull request by hehesangsj**
## Motivation

Support InternVL2.5

## Modifications

Based on PR #3351. 

Support both InternLM2ForCausalLM & Qwen2ForCausalLM as language model.

Fix bugs in multi-nodes deployment

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by zhaochenyang20** [comment]
@hehesangsj thansk, please fix the conflicts and we can merge this ASAP.

**Comment by hehesangsj** [comment]
@mickqian @zhaochenyang20 Could you review the code~

**Comment by zhaochenyang20** [comment]
> @mickqian @zhaochenyang20 Could you review the code~

hey, is this https://github.com/sgl-project/sglang/issues/4456#issuecomment-2731976868 provided by your team? Otherwise, please provide the benchmark score at #4456

Also, @yizhang2077 could you help to review? thanks so much!

File: python/sglang/srt/managers/image_processor.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
 def get_image_processor(
     hf_config, server_args: ServerArgs, processor
 ) -> BaseImageProcessor:
+    # print(IMAGE_PROCESSOR_MAPPING)
```
Comment (yizhang2077) [inline python/sglang/srt/managers/image_processor.py]: remove comment here

File: python/sglang/srt/configs/janus.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            self.cls = self.cls.__name__
+
+        self.params = kwargs.get("params", {})
+
+
+@dataclass
+class SigLIPVisionCfg:
+    width: int = 1152
+    layers: Union[Tuple[int, int, int, int], int] = 27
+    heads: int = 16
```
Comment (yizhang2077) [inline python/sglang/srt/configs/janus.py]: It seems some of config is provided in janus_pro.py ?

File: python/sglang/srt/constrained/outlines_backend.py [hunk 3]
```diff
@@ -12,7 +12,6 @@
 # limitations under the License.
 # ==============================================================================
 """Constrained decoding with outlines backend."""
-
```
Comment (yizhang2077) [inline python/sglang/srt/constrained/outlines_backend.py:15]: Do not change this file

File: python/sglang/srt/managers/schedule_batch.py [hunk 4]
```diff
@@ -139,6 +141,129 @@ def to_json(self):
...
+    ) -> List[int]:
+        """
+        Pad the input ids sequence containing media tokens, and replace them with pad_values
+        """
+        pass
+
+
+class MediaPaddingHelperTokenPairs(MediaPaddingHelper):
+    """Handles media tokens enclosed by special token pairs (e.g., <image>...</image>).
+
```
Comment (yizhang2077) [inline python/sglang/srt/managers/schedule_batch.py]: move this code into `multi_modality_padding.py`

File: python/sglang/srt/models/deepseek_janus_pro.py [hunk 5]
```diff
@@ -1984,7 +1984,6 @@ def forward(
             and len(forward_batch.image_inputs) != 0
             and forward_batch.image_inputs[0] is not None
         ):
-
```
Comment (yizhang2077) [inline python/sglang/srt/models/deepseek_janus_pro.py:1987]: Do not change here

File: python/sglang/srt/models/internlm2.py [hunk 6]
```diff
@@ -114,7 +114,6 @@ def __init__(
         self.scaling = self.head_dim**-0.5
         self.rope_theta = rope_theta
         self.max_position_embeddings = max_position_embeddings
-
```
Comment (yizhang2077) [inline python/sglang/srt/models/internlm2.py:117]: Do not change here

File: python/sglang/srt/models/internvl.py [hunk 7]
```diff
@@ -0,0 +1,733 @@
...
+    from flash_attn.bert_padding import pad_input, unpad_input
+    from flash_attn.flash_attn_interface import \
+        flash_attn_varlen_qkvpacked_func
+    has_flash_attn = True
+except:
+    print('FlashAttention2 is not installed.')
+    has_flash_attn = False
+
+
+class FlashAttention(nn.Module):
```
Comment (yizhang2077) [inline python/sglang/srt/models/internvl.py:50]: We provide some attention code in `vision.py`, could you try to use it instead of writing attention again?

File: python/sglang/srt/tokenizers/lmtokenizer.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+logger = logging.get_logger(__name__)
+
+VOCAB_FILES_NAMES = {"vocab_file": "./tokenizer.model"}
+
+PRETRAINED_VOCAB_FILES_MAP = {}
+
+
+# Modified from transformers.model.llama.tokenization_llama.LlamaTokenizer
+class InternLM2Tokenizer(PreTrainedTokenizer):
+    """
```
Comment (yizhang2077) [inline python/sglang/srt/tokenizers/lmtokenizer.py]: Why do we need extra tokenizer here? can we directly use it from transformers?

**Review by yizhang2077 — COMMENTED** [review]
Hi, I leave some simple comments here, I will keep reviewing~ and plz format your code since lint failed.

**Comment by mickqian** [comment]
@zhaochenyang20 @yizhang2077 #3351   These are duplicated

**Comment by zhaochenyang20** [comment]
> @zhaochenyang20 @yizhang2077 #3351 These are duplicated

I let them decouple and I will credit you! thanks ! @mickqian

**Comment by zhaochenyang20** [comment]
@hehesangsj fix the lint and solve the conflicts?

**Comment by zhaochenyang20** [comment]
@hehesangsj do you sovle all the reviews above? if so, please ping us and we can continue to review. thanks a lot

**Comment by hehesangsj** [comment]
> @hehesangsj do you sovle all the reviews above? if so, please ping us and we can continue to review. thanks a lot

okay, I will solve the reviews ASAP.

File: python/sglang/srt/models/internvl.py (hunk 7 above)
Comment (hehesangsj) [reply python/sglang/srt/models/internvl.py:50]: I think it's a little bit hard because the num_heads of InternViT is 25. Using `VisionAttention` can easily leads to bugs when deploying using `--tp 8`

**Comment by hehesangsj** [comment]
> @hehesangsj do you sovle all the reviews above? if so, please ping us and we can continue to review. thanks a lot

reviews are fixed and latest main branch is merged. check the code plz~

**Comment by zhaochenyang20** [comment]
@yizhang2077 yi and @mickqian could you help to review?
